src/ecrn_reg.erl

```diff
@@ -111,33 +148,43 @@ code_change(_OldVsn, State, _Extra) ->
 %%%===================================================================
 %%% Internal functions
 %%%===================================================================
-discarded_keys(Key, {discarded_keys, Keys}) ->
-    {discarded_keys, [Key | Keys]};
-discarded_keys(Key, ok) ->
-    {discarded_keys, [Key]}.
-
-add_for_keys([Key | Rest], Body, Dict, Result) ->
-    case dict:find(Key, Dict) of
-        error ->
-            add_for_keys(Rest, Body, dict:append(Key, Body, Dict), Result);
-        _Value ->
-            add_for_keys(Rest, Body, dict:append(Key, Body, Dict),
-                         discarded_keys(Key, Result))
+get_for_key(Ref, #state{r2p=Map}) when is_reference(Ref); is_atom(Ref) ->
+    case maps:find(Ref, Map) of
+        error   -> undefined;
+        OkValue -> OkValue
     end;
-add_for_keys([], _, Dict, Result) ->
-    {Result, Dict}.
-
-remove_for_keys([Key | Rest], Dict) ->
-    remove_for_keys(Rest, dict:erase(Key, Dict));
-remove_for_keys([],Dict) ->
-    Dict.
+get_for_key(Pid, #state{p2r=Map}) when is_pid(Pid) ->
+    case maps:find(Pid, Map) of
+        error   -> undefined;
+        OkValue -> OkValue
+    end.
 
-get_for_key(Key, Dict) ->
-    case dict:find(Key, Dict) of
-        error ->
-            undefined;
-        Value ->
-            Value
+find_and_remove(Ref, S = #state{r2p=M1, p2r=M2}, Fun) when is_reference(Ref); is_atom(Ref) ->
+    case maps:find(Ref, M1) of
+        {ok, Pid} ->
+            is_function(Fun, 1) andalso Fun(Pid),
+            case maps:find(Pid, M2) of
+                {ok, Refs} ->
+                    case lists:delete(Ref, Refs) of
```

Comment:
Same as above. Too much nesting one the cases

Reply:
This was refactored